### PR TITLE
Authentication processor 3/4 - Add implementation

### PR DESCRIPTION
**Pull request by jpkrohling**
**Description:** Added the OIDC authenticator. Split from #1728.

This PR is based on top of #1808. 

**Link to tracking Issue:** Part of the solution for #1424. 

**Testing:** unit tests.

**Documentation:** none.

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

### CODE REVIEW

**Comment by bogdandrutu** [comment]
Please rebase

File: internal/auth/oidc_authenticator.go [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+}
+
+func (o *oidcAuthenticator) Close() error {
+	// no-op at the moment
+	// once we implement caching of the tokens we might need this
+	return nil
+}
+
+func (o *oidcAuthenticator) UnaryInterceptor(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+	return o.unaryInterceptor(ctx, req, info, handler, o.Authenticate)
```
Comment (jpkrohling) [inline internal/auth/oidc_authenticator.go:142]: @bogdandrutu just realized that moving this to `config/configauth/internal` might be easier than I thought. I'd still do it in a follow-up PR if you agree.

**Comment by jpkrohling** [comment]
Any news on this one?

**Comment by jpkrohling** [comment]
Ping

File: internal/auth/oidc_authenticator.go [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+	o.provider = provider
+
+	o.verifier = o.provider.Verifier(&oidc.Config{
+		ClientID: o.config.Audience,
+	})
+
+	return nil
+}
+
+func (o *oidcAuthenticator) Close() error {
```
Comment (bogdandrutu) [inline internal/auth/oidc_authenticator.go:135]: In a different PR, rename to `Shutdown` and pass `Context`.

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**